---
title: "Annotation guided: U2AF eCLIP data"

output: html_notebook

---

<style>
body {
text-align: justify}
</style>


```{r initial_setup, message=F,warning=F,echo=F}

#### Load necessary libraries and variables

library(tidyverse) 
library(ggrepel) # repel text = separate overlapping text in a plot
library(ggseqlogo) # generate and plot sequence logo with ggplot
library(data.table)

library(GenomicAlignments) # Functions on aligned objects
library(rtracklayer) # interface with genome browsers (UCSC) and import/manipulation of annotation tracks
library(bamsignals) # required to generate coverage files

library(BSgenome) # genome functions
library(BSgenome.Hsapiens.UCSC.hg38) # human genome

library(edgeR) # perform differential analysis

library(extrafont) #for text in plots
loadfonts(quiet=T) 
bs<-11 #general setting in ggplot 
bf<-"Arial" #general setting in ggplot

Q157R_color<-"#5063A2"
cQ157R_color<-"#323E65"

S34F_color<-"#FC801C"
cS34F_color<-"#7E400E"

WT_color<-"#999999"
cWT_color<-"#393939"

BG_color<-"#999999"
MUT_color<-"#00aedb"

A_color<-"#0f9648"
C_color<-"#255c99"
G_color<-"#f7b32b"
U_color<-"#d62839"

load("./data/U2AF_freclip_3SS.RData") # data structures for the CLIP analysis

```



```{r useful_functions, echo=F, message=F, warning=F}

# U2AF_eCLIP functions ----

#### Custom functions to extract and work with junction data

# Function that return granges centered on intron-exon junctions (3'SS) with predefined offsets (e.g. -10;+10)
ranges_inex<-function(exon_df, # data frame/data table with exon coordinates
                      start, # column with exon start
                      end,  #column with exon end
                      chr, #column with exon chromosome (ucsc format)
                      strand, #column with exon strand (ucsc format)
                      names="", # column with names to be given to GRanges object
                      off_pre=10, # offset before inex junction
                      off_post=10,  # offset after inex junction
                      correct0base=FALSE){ # correct start for 0baseformat (0-based ucsc and 1-based ensembl) =FALSE because we will do a "manual" correction of the predefined offsets according to the strand
  off_post<-off_post-1 #off_post in 3SS is within the exon so it has to be corrected from 1-based to 0-based
  if(correct0base){
    exon_df$hl_start<-exon_df[[start]]+1 # correct for 0 base (to 1-based to use with ensembl junction coordinates)
  }else{exon_df$hl_start<-exon_df[[start]]}
  exon_df$hl_end<-exon_df[[end]] # we want to correct only the exon start coordinate because only the start is inside the 3'SS
  exon_df$gm_start<-ifelse(exon_df[[strand]]=="+",(exon_df$hl_start-off_pre),(exon_df$hl_end-off_post))
  exon_df$gm_end<-ifelse(exon_df$strand=="+",(exon_df$hl_start+off_post),(exon_df$hl_end+off_pre))
  #conventionally, the start is always smaller than the end, BUT in the negative strand the start is greater so it will be considered as the end, and the start of the inex has to be smaller than the end  
  
  out_ranges<-makeGRangesFromDataFrame(exon_df,seqnames.field = chr,start.field = "gm_start",end.field = "gm_end",strand.field=strand)
  
  if(names!=""){
    out_ranges <- setNames(out_ranges, exon_df[[names]]) #set the names on the object and returns the object
  }
  return(out_ranges)
}

# Function that return granges centered on exon-intron (5'SS) junctions with predefined offsets

ranges_exin<-function(exon_df, # data frame/data table with exon coordinates
                      start, # column with exon start
                      end,  #column with exon end
                      chr, #column with exon chromosome (ucsc format)
                      strand, #column with exon strand (ucsc format)
                      names="", # column with names to be given to GRanges object
                      off_pre=10, # offset before inex junction
                      off_post=10,  # offset after inex junction
                      correct0base=FALSE){ # correct start for 0baseformat?
  off_pre<-off_pre-1 #off_pre in 5SS is within the exon so it has to be corrected from 1-based (ensembl) to 0-based (ucsc)
  if(correct0base){
    ### exon_df$hl_end<-exon_df[,end]+1 # correct for 0 base
    exon_df$hl_end<-exon_df[[end]]+1  # correct for 0 base
    }else{exon_df$hl_end<-exon_df[[end]]}
  exon_df$hl_start<-exon_df[[start]]
  exon_df$gm_start<-ifelse(exon_df[[strand]]=="+",(exon_df$hl_end-off_pre),(exon_df$hl_start-off_post))
  exon_df$gm_end<-ifelse(exon_df$strand=="+",(exon_df$hl_end+off_post),(exon_df$hl_start+off_pre))
  
  out_ranges<-makeGRangesFromDataFrame(exon_df,seqnames.field = chr,start.field = "gm_start",end.field = "gm_end",strand.field=strand)
  
  if(names!=""){
    out_ranges <- setNames(out_ranges, exon_df[[names]])
  }
  return(out_ranges)
}


# Function that extract RNAStringSet from a granges object and a genome

RNAsequences<-function(in_ranges, # granges object
                       genome=BSgenome.Hsapiens.UCSC.hg38, # genome to be used
                       u_conv=TRUE){ # convert the DNA sequence to RNA: (T to U)
  
  DNA<-getSeq(genome,in_ranges)
  ifelse(u_conv, RNA<-RNAStringSet(gsub("T","U",as.character(DNA))), RNA<-DNA)
  return(RNA)
}


# Function to extract coverage given a set of junctions (used for metaprofiles)

jCov<-function(input_list, # list with the coverage values of each junction (x3_list)
               junction_vec,  #vector of junction names (defined by coordinates)
               name="set1"){        # name given to the analysis (in a separate column)
  out_dt<-NULL
  for(exp in names(input_list)){
    exp_dt<-input_list[[exp]]
    exp_dt<-exp_dt[id %in%junction_vec]
    
    filt_vec_dt<-data.table("id"=exp_dt$id, "rowSum"=rowSums(exp_dt[,-"id"]))
    filt_vec_dt<-filt_vec_dt[rowSum<=2^boxplot(log2(rowSum[rowSum>0]),plot = F)$stats[5]]
    filt_dt<-exp_dt[id %in% filt_vec_dt$id] 
    
    # the boxplot formula calculates 5 thresholds, three for the definition of the box, two for the whiskers. 
    # the 5th threshold is the upper whisker and we want to eliminate junctions with "excess" of signal
    # 2^ because we want to remove the log2 (applied by default in the boxplot)
    exp_out<-data.table(position=as.integer(colnames(exp_dt[,-"id"])), # positions relative to 3'SS
                        n_all=nrow(exp_dt), # number of all junctions considered
                        n_trim=nrow(filt_dt), # number of junctions (trimmed for outliers)
                        sum_all=colSums(exp_dt[,-"id"]), # sum of the signal (all)
                        sum_trim=colSums(filt_dt[,-"id"]), # sum of the signal (trimmed)
                        mean_all=colMeans(exp_dt[,-"id"]), # mean of the signal (all)
                        mean_trim=colMeans(filt_dt[,-"id"]), # mean of the signal (trimmed)
                        sample=exp, stringsAsFactors = F) # sample
    out_dt<-rbind(out_dt,exp_out)
  }
  out_dt$jclass<-name
  return(out_dt)
}

# Function to build datatable/frames for plotting profiles of selected junctions

jBuild <- function(sample_dt){ #dt filtered with conditions to consider
  
  t_test_dt<-unique(sample_dt[,t.test(signal_per_million ~ genotype,.SD), by=c("wrapper","position")], by=c("wrapper","position","p.value"))[,p.value_1tail:=p.value/2]

  cond_dt <- sample_dt[,.(N=.N, median=median(signal_per_million),
                          mean=mean(signal_per_million), 
                          sd=sd(signal_per_million)), by=c("protein","genotype","treatment","fraction","library","condition","jclass","position","wrapper")][
                            ,se:=sd/sqrt(N)]
  
  cond_dt[is.na(se), se:= mean*0.1]
  
  plot_roof<- cond_dt[,.(plot_roof=max(mean),consistency_threshold=max(mean-se)), by="wrapper"][,
                       plot_validation:=consistency_threshold>plot_roof/20][,c("wrapper","plot_roof","plot_validation")]
                                                            
  cond_dt <- cond_dt[plot_roof,on="wrapper"]

  cond_dt <- cond_dt[,point_validation:=(mean-se)>(plot_roof/20)]
  
  ref_cond_df <- cond_df %>% dplyr::filter(genotype=="WT") %>%
    dplyr::select(position,wrapper,mean,se)
  colnames(ref_cond_df) <- c("position","wrapper","mean_WT","se_WT")
  
  ref_cond_dt <- cond_dt[genotype=="WT",c("position","wrapper","mean","se")]
  colnames(ref_cond_dt) <- c("position","wrapper","mean_WT","se_WT")
  
  delta_dt <- cond_dt[ref_cond_dt, on=c("position","wrapper")]
  delta_dt <- delta_dt[,delta_mean:=mean-mean_WT][, delta_se:=se+se_WT][genotype!="WT"]
  delta_dt <- delta_dt[t_test_dt, on=c("wrapper","position")]
  return(list("cond_dt"=cond_dt,"delta_dt"=delta_dt))
}


## Crosslink coverage extraction cycle  from BAM files ----

clip_run<-function(bam_dir, x_ranges){
  
  list_input<-list.files(path=bam_dir, pattern="bam$", full.names=F) #bam names with ""
  x3_ranges<-x_ranges
  x3_list<-list()

  for (bam_file in list_input){
  
    bam_name<-substr(bam_file,1,(nchar(bam_file)-6))
    trial_3<-bamProfile(bampath=paste0(bam_dir,bam_file),x3_ranges,verbose=F,ss=T) # consider the strand
    sense_3<-lapply(trial_3, function(x) x[1,]) # keep sense signal
    df_struct_3<-(as.data.frame(do.call(rbind, as.list(sense_3))))
    colnames(df_struct_3)<-c(seq(-40,-1),seq(1,10)) # crosslinking shifts of -1, and deletion of 0
    rownames(df_struct_3)<-names(x3_ranges)
    x3_list[[bam_name]]<-df_struct_3
  }

  return(x3_list)  
  
}

```

### Annotation of the available samples

```{r sample_annotation, echo=FALSE,message=FALSE,warning=FALSE}

c_anno<-data.table("sample"=colnames(junct_cov_dt[,-"id"]))
c_anno_add <- as.data.table(c_anno)[,c("protein","genotype","treatment","fraction","library","replicate") := tstrsplit(sample,"_",fixed=T)][,-"sample"]
c_anno<-cbind(c_anno,c_anno_add)
c_anno[,condition:=paste(protein,genotype,fraction,sep="_")]
c_anno[, .(number=.N), by=c("protein", "genotype","fraction")]
```

### Generation of an alluvial plot describing 3'SS junctions

```{r alluvial_trial, message=F,warning=F,fig.height=4,fig.width=8}

library(ggalluvial)
coord_dt<-coord_dt[,AG:=str_sub(inex_RNA,39,40)][,m3:=str_sub(inex_RNA,38,38)][,p1:=str_sub(inex_RNA,41,41)]

setkey(exon_anno_dt,exon_chrom_start, exon_chrom_end, chromosome_ucsc, strand_ucsc, coord)
setkey(coord_dt,exon_chrom_start, exon_chrom_end, chromosome_ucsc, strand_ucsc, coord)

allu_pre_dt<-exon_anno_dt[coord_dt, on=c("exon_chrom_start", "exon_chrom_end", "chromosome_ucsc", "strand_ucsc", "coord"), nomatch=0] # inner join

allu_pre_dt[which(AG!="AG"),"AG"]<-"non AG"
allu_pre_dt[which(AG!="AG"),"m3"]<-"non AG"
allu_pre_dt[which(gene_biotype!="protein_coding"),"gene_biotype"]<-"non_coding"

allu_dt<-unique(allu_pre_dt[exon_class!="U",c("exon_class","AG","m3","gene_biotype","coord")])[,.(freq= .N), by=c("exon_class","AG","m3","gene_biotype")]

allu_dt[,exon_class:=factor(exon_class, levels=c("F","I","L"))]
allu_dt[,exon_class:=recode(exon_class, "F" = "First","I" = "Internal","L" = "Last")]

allu_dt[,AG:=factor(AG,levels=c("non AG","AG"))]
allu_dt[,m3:=factor(m3,levels=c("non AG","C","U","A","G"))]

allu_dt <- allu_dt[order(exon_class,AG,m3)]

allu_plot<-ggplot(allu_dt,
       aes(y = freq,
           axis1 = exon_class, axis2 = gene_biotype, axis3 = AG)) +
  geom_alluvium(aes(fill = m3), colour="white",
                width = 0, knot.pos = 1/6, na.rm=T,reverse = FALSE) +
  theme_void()+
  theme(legend.position ="top")+
  scale_fill_manual(values=c(WT_color,C_color,U_color,A_color,G_color))+
  #guides(fill = FALSE) +
  geom_stratum(width = 1/7, reverse = FALSE, colour="white",fill="grey60") +
  geom_text(stat = "stratum", infer.label = TRUE, reverse = FALSE,colour="white",fontface=2) +
  scale_x_continuous(breaks = 1:3, labels = c("m3", "AG", "Class")) +
  coord_flip()

allu_plot

```

### Generation of the U2AF freCLIP metaprofile around covered 3'SS junctions

```{r metaprofile frac all,fig.height=5,fig.width=5,echo=F,message=F}


junc_sele <- unique(exon_anno_dt$coord)

out3<-jCov(input_list=x3_list_dt,
           junction_vec=junc_sele,
           name="meta_3SS")

setkey(c_anno,sample)
input_dt <- out3[c_anno, on="sample"]

#consider each SINGLE SAMPLE and calculate the total signal from all the positions and the percentage of the signal for each position 

sample_meta_dt<-input_dt[, c("tot_trim","tot_all"):=list(sum(sum_trim), sum(sum_all)), by= c("sample","jclass")][,signal_perc_trim:=(as.numeric(sum_trim)/as.numeric(tot_trim))*100][,signal_perc_all:= (as.numeric(sum_all)/as.numeric(tot_all))*100]
#or _all

#group all except replicates to calculate the stats and to plot the mean signal from the replicates

cond_meta_dt<-sample_meta_dt[, .(N=.N,median=median(signal_perc_trim),mean=mean(signal_perc_trim),sd=sd(signal_perc_trim)), .(protein,genotype,treatment,fraction,library,condition,jclass,position)][, se := sd/sqrt(N)]        
cond_meta_dt[is.na(se), se:= mean*0.1]

input_selection<-cond_meta_dt[position%in%seq(-20,5)] # select specific nucleotides for the metaprofile
input_selection[, genotype:=factor(genotype,levels=c("WT","S34F"))]

input_selection[which(condition %in% c("U2AF1_WT_heavy","U2AF1_S34F_heavy")),"condition"] <- "Heavy fraction (U2AF1+U2AF2)"

input_selection[which(condition %in% c("U2AF1_WT_light","U2AF1_S34F_light")),"condition"] <- "Light fraction (U2AF1)"

input_selection[,condition:=factor(condition,levels=c("Heavy fraction (U2AF1+U2AF2)","Light fraction (U2AF1)"))]

input_selection<-input_selection[order(genotype, position)]
bs<-11
ex<-ggplot(input_selection,aes(position,mean,col=genotype,fill=genotype)) +
    geom_vline(xintercept=0,linetype=2,colour="grey20",alpha=0.8) +
    facet_wrap(~ condition,scales="free_y",ncol=1) +
    geom_ribbon(aes(ymin=mean-se,ymax=mean+se),alpha=0.2,colour=NA)+
    geom_line(alpha=0.7) +
    geom_point(alpha=0.8, size=bs*0.075) +
    theme_bw(base_size = bs,base_family = bf)+
    scale_colour_manual(name="genotype", values=c(WT_color,U_color)) +
    scale_fill_manual(name="genotype", values=c(WT_color,U_color)) +
    scale_x_continuous(breaks=c(seq(-100,-5,5),seq(5,100,5)))+
    scale_y_continuous(breaks=seq(0,100,3))+
    theme(legend.justification=c(0.01,0.99), legend.position=c(0.01,0.99)) +
    theme(plot.title = element_text(hjust = 0.5,size=0.75*bs))+
    theme(legend.text = element_text(size = 0.7*bs), legend.key.size = unit((0.015*bs),"in"))+
    theme(legend.title = element_blank())+
    theme(legend.box = "none")+
    theme(panel.grid.minor = element_blank())+
    labs(x="distance from junction (nt)",y="X-link events (%)")
ex

#ggsave("./3SS_metaprofile.pdf",ex,device=cairo_pdf,width = 3.5, height = 3.8, units = c("in"))

```

### Generate a logo from the sequences  of 3'SS junctions

```{r plot_logos, message=F,warning=F,echo=F, fig.height=3,fig.width=10}

# retrieve the range and the sequence used for the 3`SS metaprofile (50 nucleotides, -40 to +10)
inex_ranges<-ranges_inex(exon_df=coord_dt,
                         start="exon_chrom_start",
                         end="exon_chrom_end",
                         chr="chromosome_ucsc",
                         strand="strand_ucsc",
                         names="coord",
                         off_pre=40,
                         off_post=10,
                         correct0base=FALSE) 

inex_RNA<-RNAsequences(in_ranges=inex_ranges,
                       genome=BSgenome.Hsapiens.UCSC.hg38)

inexLogo<-ggseqlogo(as.character(inex_RNA),seq_type="rna",method="bits")

inexLogo

```

### Generate a clip profile from a single 3'SS junction

```{r metaprofile_poster,fig.height=6,fig.width=7,echo=F,message=F}


out3<-jCov(input_list=x3_list_dt,
           junction_vec="128809157_128809204_chr11_+",
           name="internal")


setkey(c_anno,sample)
input_dt <- out3[c_anno, on="sample"]


#consider each SINGLE SAMPLE and calculate the total signal from all the positions and the percentage of the signal for each position 
sample_meta_dt<-input_dt[, c("tot_trim","tot_all"):=list(sum(sum_trim), sum(sum_all)), by= c("sample","jclass")][,signal_perc_trim:=(as.numeric(sum_trim)/as.numeric(tot_trim))*100][,signal_perc_all:= (as.numeric(sum_all)/as.numeric(tot_all))*100]

cond_meta_dt<-sample_meta_dt[, .(N=.N,median=median(signal_perc_all),mean=mean(signal_perc_all),sd=sd(signal_perc_all)), .(protein,genotype,treatment,fraction,library,condition,jclass,position)][, se := sd/sqrt(N)]        
cond_meta_dt[is.na(se), se:= mean*0.1]


input_selection<-cond_meta_dt[position%in%seq(-20,5)]
input_selection[, genotype:=factor(genotype,levels=c("WT","S34F"))]

input_selection[which(condition %in% c("U2AF1_WT_heavy","U2AF1_S34F_heavy")),"condition"] <- "Heavy fraction (U2AF1+U2AF2)"
input_selection[which(condition %in% c("U2AF1_WT_light","U2AF1_S34F_light")),"condition"] <- "Light fraction (U2AF1)"

input_selection[,condition:=factor(condition,levels=c("Heavy fraction (U2AF1+U2AF2)","Light fraction (U2AF1)"))]

plot_roof <- input_selection[,.(plot_roof=max(mean+se)), .(condition)]

input_selection<-input_selection[plot_roof, on="condition"]

j_seq_dt<-data.table("position"=c(seq(-20,-1),seq(1,5)), "nucleotide"=str_split(as.character(subseq(inex_RNA["128809157_128809204_chr11_+"],21,45)),"")[[1]], "n_col"="grey30", stringsAsFactors = FALSE)

j_seq_dt[which(nucleotide=="U"),"n_col"]<- U_color
j_seq_dt[which(nucleotide=="A"),"n_col"]<- A_color
j_seq_dt[which(nucleotide=="C"),"n_col"]<- C_color
j_seq_dt[which(nucleotide=="G"),"n_col"]<- G_color

j_seq_dt<-unique(input_selection[,c("position","condition","plot_roof")])[j_seq_dt, on="position"]

j_seq_dt[,genotype:="WT"]

bs<-11
ex<-ggplot(input_selection,aes(position,mean,col=genotype,fill=genotype)) +
    geom_vline(xintercept=0,linetype=2,colour="grey20",alpha=0.8) +
    #facet_grid(genexp ~ jclass,scales = "free") +
    facet_wrap(~ condition,scales="free_y",ncol=1) +
    geom_ribbon(aes(ymin=mean-se,ymax=mean+se),alpha=0.2,colour=NA)+
    geom_line(alpha=0.8) +
    geom_point(alpha=0.9, size=bs*0.075) +
    theme_bw(base_size = bs,base_family = bf)+
    scale_colour_manual(name="genotype", values=c(WT_color,U_color)) +
    scale_fill_manual(name="genotype", values=c(WT_color,U_color)) +
    scale_x_continuous(breaks=c(seq(-100,-5,5),seq(5,100,5)))+
    #scale_y_continuous(breaks=seq(0,100,3))+
    theme(legend.justification=c(0.01,0.99), legend.position=c(0.01,0.99)) +
    theme(plot.title = element_text(hjust = 0.5,size=0.75*bs))+
    theme(legend.text = element_text(size = 0.7*bs), legend.key.size = unit((0.015*bs),"in"))+
    theme(legend.position = "none")+
    theme(legend.title = element_blank())+
    theme(legend.box = "none")+
    theme(panel.grid.minor = element_blank())+
    labs(x="distance from junction (nt)",y="X-link events (%)",title="FLI1-205; exon 8; \n chr11:128809157-128809204+")+
    geom_text(data=j_seq_dt, aes(y=-(plot_roof/10), label=nucleotide),colour=j_seq_dt$n_col,size=0.256*bs)
ex

# ggsave("./3SS_single.pdf",ex,device=cairo_pdf,width = 3.5, height = 3.8, units = c("in"))

```

